backend/uclapi/oauth/tests.py

```diff
@@ -0,0 +1,191 @@
+        self.factory = APIRequestFactory()
+
+    def test_decorator_no_client_secret_proof_provided(self):
+        request = self.factory.get('/')
+        response = self.dec_view(request)
+
+        content = json.loads(response.content.decode())
+
+        self.assertEqual(response.status_code, 400)
+        self.assertFalse(content["ok"])
+        self.assertEqual(content["error"], "No Client Secret Proof provided")
+
+    def test_decorator_no_token_provided(self):
+        request = self.factory.get(
+            '/',
+            {'client_secret_proof': 'not_a_real_proof'}
+        )
+        response = self.dec_view(request)
+
+        content = json.loads(response.content.decode())
+
+        self.assertEqual(response.status_code, 400)
+        self.assertFalse(content["ok"])
+        self.assertEqual(content["error"], "No token provided via GET.")
+
+    def test_decorator_nonexistent_token_provided(self):
+        request = self.factory.get(
+            '/',
+            {'client_secret_proof': 'not_a_real_secret_proof', 'token': 'fake'}
+        )
+        response = self.dec_view(request)
+
+        content = json.loads(response.content.decode())
+
+        self.assertEqual(response.status_code, 400)
+        self.assertFalse(content["ok"])
+        self.assertEqual(content["error"], "Token does not exist")
+
+    def test_decorator_client_secret_verification_failed(self):
+        # create User, App, and OAuth it
+        user_ = User.objects.create(
+            email="test@ucl.ac.uk", cn="test",
+            given_name="Test Test"
+        )
+        app_ = App.objects.create(user=user_, name="An App")
+        oauth_scope = OAuthScope.objects.create()
+        oauth_token = OAuthToken.objects.create(
+            app=app_, user=user_, scope=oauth_scope
+        )
+
+        request = self.factory.get(
+            '/',
+            {
+                'client_secret_proof': 'not_a_real_proof',
+                'token': oauth_token.token
+            }
+        )
+        response = self.dec_view(request)
+
+        content = json.loads(response.content.decode())
+
+        self.assertEqual(response.status_code, 400)
+        self.assertFalse(content["ok"])
+        self.assertEqual(
+            content["error"],
+            "Client secret HMAC verification failed."
+        )
+
+    def test_decorator_no_permission_to_access(self):
+        pass
+
+    def test_decorator_everything_passes(self):
+        user_ = User.objects.create(
+            email="test@ucl.ac.uk", cn="test",
+            given_name="Test Test"
+        )
+        app_ = App.objects.create(user=user_, name="An App")
+        oauth_scope = OAuthScope.objects.create()
+        oauth_token = OAuthToken.objects.create(
+            app=app_, user=user_, scope=oauth_scope
+        )
+
+        # TODO: THIS WILL GET CHANGED
+        # calculate the client_secret_proof
+        import hmac, hashlib, base64
```

Comment:
@ChristopherHammond13 I had no idea how to get / obtain / calculate `client_secret_proof` so I calculated it like this - yes or no?